### PR TITLE
Implement JSON array filtering functions

### PR DESCRIPTION
**Pull request by Jinxiao0302**
Related to  #1410

**Technical details**
Added a DBFunction for getting JSON array length;
To be done: Add comparaing functions that nest array length function.

## Checklist
 
- [ ] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `master` branch of the repository
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: db/functions/hints.py
Comment (dmos62) [inline db/functions/hints.py:131]: You'll want to tell Mathesar how these hints relate to actual types. Look at `db/types/hintsets.py`. Note this line:

```
    _add_to_db_type_hintsets(categories.JSON_TYPES, (hints.json,))
```

This says that all database types defined as `categories.JSON_TYPES` are associated with the hints `(hints.json,)`.

You'll probably want to make changes to this. And, to the related categories in the `db.types.categories` as as well, since currently `categories.JSON_TYPES` is a catch-all category for everything JSON, and you probably want something more granular.

Note how the `categories` sets define a sort of hierarchy, where for example time-related types are made up of point-in-time as well as duration types, those in turn are made up of other more granular categories, etc. You might want something like this for JSON, though I wouldn't invest too much time into that right now: the mentioned hierarchy was borne out of necessity.

By the way, if this wasn't clear before (probably not), the way Mathesar finds what hints are associated with what UI type, is it takes the intersection of the UI type's DB types' hints (a UI type is effectively a set of DB types, and each of those have a set of hints, as seen in `db/types/hintsets.py`); and, the way it checks whether a filter is applyable to a column of a given UI type, is that it checks if the hints of the first parameter of that filter are satisfied by the hints of the UI type. "Satisfied" here is intentionally ambiguous, since we want the freedom to evolve how this works, but at the moment this mostly means that the first param hint-set must be a subset of the UI type hint-set.

Oof, sorry for the wall of text. Thumbs up for braving through :+1::+1:, feel free to reach out.

File: db/functions/base.py
Comment (dmos62) [inline db/functions/base.py:317]: This isn't actually a filter. You'll probably use this function to compose a `DBFunctionPacked`-based filter and you'll give that a `mathesar_filter` hint, but giving this function here a `mathesar_filter` will result in unexpected behaviour.

**Review by dmos62 — COMMENTED** [review]
Regarding your question about how to use the `ArrayLength` to compose a filter like `json array length is less than` (I presume that that's a valid approximation of your goal), create a `DBFunctionPacked` subclass inside `db.types.custom.json_array`. `DBFunctionPacked` (and the whole of `hints.mathesar_filter` mechanism, really) is there to bypass the lack of composition in our current UI filtering system.

File: db/functions/base.py
Comment (Jinxiao0302) [reply db/functions/base.py]: Thanks for your very helpful explanation! @dmos62 , I added a DBFunctionPacked function using ArrayLength in the `db/types/custom/json_array` and also moved the ArrayLength into that file. Also, I added the json array into categories.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Hey @Jinxiao0302 , as noted through other channels I'd like you to add some tests before we merge this PR. Thank you, it's looking good so far!

**Review by pavish — APPROVED** [review]
The changes look good to me. As @mathemancer mentioned, we can merge this in once we have tests. I'm approving from my end.

**Review by mathemancer — APPROVED** [review]
Okay, I think this looks good. Let's get it merged!

**Pull request merged**
